# ADO (Accrued Days Off) Roster (MMC)

Rostering at MMC is dependant on individuals and their skill sets.  However, in the interest of anonymity we will use individuals initials instead of names.

## General Medicine

General Medicine is relatively simpler than the other four teams as it does not have a skill set requisite - all team members have the same skills.

### Introduction

We have 9 full-time pharmacists in the Gen Med team.  Let each pharmacist be $p$, where $p$ ranges from [1,9].

We will make values of $p$ strings, with the values:

|$p$|string name|
|:-:|:-:|
|1|KA|
|2|AA|
|3|JA|
|4|ST|
|5|TA|
|6|SY|
|7|JD|
|8|CA|
|9|SA|
|10|AN|
|11|JP|

In 2018, we want to ascribe each full-time pharmacist an ADO once every four weeks.  We have 250 normal working days in 2018 on which an ADO can occur.  We let these days be $d$, where $d$ ranges from [1,250].


The string values for $d$ can be generated here to prepare for CPLEX input

In [77]:
[str(i).zfill(3) for i in range(1,251)]

['001',
 '002',
 '003',
 '004',
 '005',
 '006',
 '007',
 '008',
 '009',
 '010',
 '011',
 '012',
 '013',
 '014',
 '015',
 '016',
 '017',
 '018',
 '019',
 '020',
 '021',
 '022',
 '023',
 '024',
 '025',
 '026',
 '027',
 '028',
 '029',
 '030',
 '031',
 '032',
 '033',
 '034',
 '035',
 '036',
 '037',
 '038',
 '039',
 '040',
 '041',
 '042',
 '043',
 '044',
 '045',
 '046',
 '047',
 '048',
 '049',
 '050',
 '051',
 '052',
 '053',
 '054',
 '055',
 '056',
 '057',
 '058',
 '059',
 '060',
 '061',
 '062',
 '063',
 '064',
 '065',
 '066',
 '067',
 '068',
 '069',
 '070',
 '071',
 '072',
 '073',
 '074',
 '075',
 '076',
 '077',
 '078',
 '079',
 '080',
 '081',
 '082',
 '083',
 '084',
 '085',
 '086',
 '087',
 '088',
 '089',
 '090',
 '091',
 '092',
 '093',
 '094',
 '095',
 '096',
 '097',
 '098',
 '099',
 '100',
 '101',
 '102',
 '103',
 '104',
 '105',
 '106',
 '107',
 '108',
 '109',
 '110',
 '111',
 '112',
 '113',
 '114',
 '115',
 '116',
 '117',
 '118',
 '119',
 '120',
 '121',
 '122',
 '123',
 '124',
 '125',


In [4]:
import numpy as np
import pandas as pd
import os
os.listdir("Current Rosters/")

['2017_Critical Care Team.xlsx',
 '2017_Digestive Health_Nephrology Daily Roster.xlsx',
 '2017_General Medicine Team.xlsx',
 '2017_General_Medicine_Team.csv',
 '2017_Special Medicine Team.xlsx',
 'All team Rosters']

In [26]:
df = pd.read_table('Current Rosters/2017_General_Medicine_Team.csv', sep = ",", header = 0)

In [30]:
pd.set_option('display.max_rows', 260)
df

d        date        day
0      1   2/01/2019  Wednesday
1      2   3/01/2019   Thursday
2      3   4/01/2019     Friday
3      4   7/01/2019     Monday
4      5   8/01/2019    Tuesday
5      6   9/01/2019  Wednesday
6      7  10/01/2019   Thursday
7      8  11/01/2019     Friday
8      9  14/01/2019     Monday
9     10  15/01/2019    Tuesday
10    11  16/01/2019  Wednesday
11    12  17/01/2019   Thursday
12    13  18/01/2019     Friday
13    14  21/01/2019     Monday
14    15  22/01/2019    Tuesday
15    16  23/01/2019  Wednesday
16    17  24/01/2019   Thursday
17    18  25/01/2019     Friday
18    19  29/01/2019    Tuesday
19    20  30/01/2019  Wednesday
20    21  31/01/2019   Thursday
21    22   1/02/2019     Friday
22    23   4/02/2019     Monday
23    24   5/02/2019    Tuesday
24    25   6/02/2019  Wednesday
25    26   7/02/2019   Thursday
26    27   8/02/2019     Friday
27    28  11/02/2019     Monday
28    29  12/02/2019    Tuesday
29    30  13/02/2019  Wednesday
30    31  14/02/2019   Thursday
31    32  15/02/2019     Friday
32    33  18/02/2019     Monday
33    34  19/02/2019    Tuesday
34    35  20/02/2019  Wednesday
35    36  21/02/2019   Thursday
36    37  22/02/2019     Friday
37    38  25/02/2019     Monday
38    39  26/02/2019    Tuesday
39    40  27/02/2019  Wednesday
40    41  28/02/2019   Thursday
41    42   1/03/2019     Friday
42    43   4/03/2019     Monday
43    44   5/03/2019    Tuesday
44    45   6/03/2019  Wednesday
45    46   7/03/2019   Thursday
46    47   8/03/2019     Friday
47    48  12/03/2019    Tuesday
48    49  13/03/2019  Wednesday
49    50  14/03/2019   Thursday
50    51  15/03/2019     Friday
51    52  18/03/2019     Monday
52    53  19/03/2019    Tuesday
53    54  20/03/2019  Wednesday
54    55  21/03/2019   Thursday
55    56  22/03/2019     Friday
56    57  25/03/2019     Monday
57    58  26/03/2019    Tuesday
58    59  27/03/2019  Wednesday
59    60  28/03/2019   Thursday
60    61  29/03/2019     Friday
61    62   1/04/2019     Monday
62    63   2/04/2019    Tuesday
63    64   3/04/2019  Wednesday
64    65   4/04/2019   Thursday
65    66   5/04/2019     Friday
66    67   8/04/2019     Monday
67    68   9/04/2019    Tuesday
68    69  10/04/2019  Wednesday
69    70  11/04/2019   Thursday
70    71  12/04/2019     Friday
71    72  15/04/2019     Monday
72    73  16/04/2019    Tuesday
73    74  17/04/2019  Wednesday
74    75  18/04/2019   Thursday
75    76  23/04/2019    Tuesday
76    77  24/04/2019  Wednesday
77    78  26/04/2019     Friday
78    79  29/04/2019     Monday
79    80  30/04/2019    Tuesday
80    81   1/05/2019  Wednesday
81    82   2/05/2019   Thursday
82    83   3/05/2019     Friday
83    84   6/05/2019     Monday
84    85   7/05/2019    Tuesday
85    86   8/05/2019  Wednesday
86    87   9/05/2019   Thursday
87    88  10/05/2019     Friday
88    89  13/05/2019     Monday
89    90  14/05/2019    Tuesday
90    91  15/05/2019  Wednesday
91    92  16/05/2019   Thursday
92    93  17/05/2019     Friday
93    94  20/05/2019     Monday
94    95  21/05/2019    Tuesday
95    96  22/05/2019  Wednesday
96    97  23/05/2019   Thursday
97    98  24/05/2019     Friday
98    99  27/05/2019     Monday
99   100  28/05/2019    Tuesday
100  101  29/05/2019  Wednesday
101  102  30/05/2019   Thursday
102  103  31/05/2019     Friday
103  104   3/06/2019     Monday
104  105   4/06/2019    Tuesday
105  106   5/06/2019  Wednesday
106  107   6/06/2019   Thursday
107  108   7/06/2019     Friday
108  109  11/06/2019    Tuesday
109  110  12/06/2019  Wednesday
110  111  13/06/2019   Thursday
111  112  14/06/2019     Friday
112  113  17/06/2019     Monday
113  114  18/06/2019    Tuesday
114  115  19/06/2019  Wednesday
115  116  20/06/2019   Thursday
116  117  21/06/2019     Friday
117  118  24/06/2019     Monday
118  119  25/06/2019    Tuesday
119  120  26/06/2019  Wednesday
120  121  27/06/2019   Thursday
121  122  28/06/2019     Friday
122  123   1/07/2019     Monday
123  124   2/07/2019    Tuesday
124  12

### Decision Variables

We have that:

$$y_{pd}$$ for some pharmacist $y$ that has an ADO on day $d$

### Objective Function

### Constraints

_One ADO each per month_

$$\sum_d \; y_{pd} \quad \forall p$$

_Birthday Constraint_

If a pharmacist has a birthday on a weekday, they receive an ADO on that day.

As this data is not available to us, we will just make up imaginary birthdays for these people.

|Pharmacist|Birthday|$d$|
|:-:|:-:|:-:|
|KA|22/04|77|
|AA|04/03|43|
|JA|11/09|175|
|ST|11/11|216|
|TA|23/03|-|
|SY|25/05|-|
|JD|01/07|123|
|CA|10/08|-|
|SA|03/09|169|

Therefore, we have the following:

$$y_{KA,77} = 1$$
$$y_{AA,43} = 1$$
$$y_{JA,175} = 1$$
$$y_{ST,216} = 1$$
$$y_{JD,123} = 1$$
$$y_{SA,169} = 1$$


_Annual Leave Constraint_

Pharmacists have pre-approved leave that also needs to be honoured. We will use the actual leave constraints: